pkg/gatewaysecret/handler.go

```diff
@@ -0,0 +1,137 @@
+package gatewaysecret
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	certmanagerv1 "github.com/cert-manager/cert-manager/pkg/apis/certmanager/v1"
+	"github.com/go-logr/logr"
+	apicorev1 "k8s.io/api/core/v1"
+	apimetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/kyma-project/lifecycle-manager/internal/pkg/flags"
+)
+
+const (
+	GatewaySecretName = "istio-gateway-secret"
+	kcpRootSecretName = "klm-watcher"
+	kcpCACertName     = "klm-watcher-serving"
+	istioNamespace    = flags.DefaultIstioNamespace
+)
+
+type GatewaySecretHandler struct {
+	*secretManager
+}
+
+func NewGatewaySecretHandler(kcpClient client.Client) *GatewaySecretHandler {
+	return &GatewaySecretHandler{
+		secretManager: &secretManager{
+			kcpClient: kcpClient,
+		},
+	}
+}
+
+func (gsh *GatewaySecretHandler) ManageGatewaySecret(rootSecret *apicorev1.Secret) error {
+	gwSecret, err := gsh.findGatewaySecret()
+
+	if isNotFound(err) {
+		return gsh.handleNonExisting(rootSecret)
+	}
+	if err != nil {
+		return err
+	}
+
+	return gsh.handleExisting(rootSecret, gwSecret)
+}
+
+func (gsh *GatewaySecretHandler) handleNonExisting(rootSecret *apicorev1.Secret) error {
+	gwSecret := gsh.newGatewaySecret(rootSecret)
+	return gsh.create(context.Background(), gwSecret)
+}
+
+func (gsh *GatewaySecretHandler) handleExisting(rootSecret *apicorev1.Secret, gwSecret *apicorev1.Secret) error {
+	caCert := certmanagerv1.Certificate{}
+	if err := gsh.kcpClient.Get(context.Background(),
```

Comment:
we should not use Background context, check the function comment, this context will never cancelled.